pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,385 @@
+	shard, ok := obj.(*tenancyv1alpha1.WorkspaceShard)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("got %T when handling deleted WorkspaceShard", obj))
+		return
+	}
+	klog.Infof("handling removed shard %q", shard.Name)
+	workspaces, err := c.workspaceIndexer.ByIndex(targetShardName, shard.Name)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	for _, workspace := range workspaces {
+		key, err := cache.MetaNamespaceKeyFunc(workspace)
+		if err != nil {
+			runtime.HandleError(err)
+			return
+		}
+		klog.Infof("queuing orphaned workspace %q", key)
+		c.queue.AddRateLimited(key)
+	}
+}
+
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting Workspace controller")
+	defer klog.Info("Shutting down Workspace controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	klog.Infof("processing key %q", key)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	err := c.process(ctx, key)
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		klog.Infof("Successfully reconciled %q", key)
+		c.queue.Forget(key)
+		return
+	}
+
+	// Re-enqueue up to 5 times.
+	num := c.queue.NumRequeues(key)
+	if errors.IsRetryable(err) || num < 5 {
+		klog.Errorf("Error reconciling key %q, retrying... (#%d): %v", key, num, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	// Give up and report error elsewhere.
+	c.queue.Forget(key)
+	runtime.HandleError(err)
+	klog.Errorf("Dropping key %q after failed retries: %v", key, err)
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	namespace, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != "" {
+		return fmt.Errorf("namespace %q found in key for cluster-wide Workspace object", namespace)
+	}
+	clusterName, _ := clusters.SplitClusterAwareKey(clusterAwareName)
+
+	obj, err := c.workspaceLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		return err
+	}
+	previous := obj.DeepCopy()
+
+	if err := c.reconcile(ctx, obj); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous.Spec, obj.Spec) {
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().Workspaces().Update(ctx, obj, metav1.UpdateOptions{})
```

Comment:
Update is ok for now, but this will need to be a Patch call to avoid and older client accidentally dropping fields added in a new revision of the CRD (same apiversion, fields added).